### DISCUSSION TITLE
Getting remark-gfm and remark-lint-no-auto-link-without-protocol to play nicely together

### DISCUSSION THREAD

**Discussion by Trott** [remarkjs/remark#863] [Q&A] [answered]
I'm using `remark` to format and lint my markdown. I've run into an issue where `remark-gfm` generates code that is flagged by `remark-lint-no-auto-link-without-protocol` and I'm trying to figure out if there is a way to get my ideal imagined possibility.

## What I'm seeing

`input.md`:

```markdown
Here is an email address to autolink: tituswormer@example.com
```

```console
$ remark -u gfm input.md -o output.md
input.md > output.md: written
$ cat output.md
Here is an email address to autolink: <tituswormer@example.com>
$ remark -u preset-lint-recommended input.md 
Here is an email address to autolink: tituswormer@example.com
input.md: no issues found
$ remark -u preset-lint-recommended output.md  
Here is an email address to autolink: <tituswormer@example.com>
output.md
  1:39-1:64  warning  All automatic links must start with a protocol  no-auto-link-without-protocol  remark-lint

⚠ 1 warning
$
```

## What I want to happen instead

I would like for `remark-gfm` output to be acceptable to `remark-lint-no-auto-link-without-protocol` without changing the text that is seen by the end user. Either of the following would be ideal for me (as far as I can tell at the moment):

* I would like for `remark-lint-no-auto-link-without-protocol` to require protocols for non-email auto links but to leave email links alone. No protocol needed on email links. In GitHub and in VS Code, markdown with `<foo@example.com>` is rendered as a `mailto:` link, and that's good enough for me.
* I would like for `remark-gfm` to insert the `mailto` protocol on email autolinks it generates but to leave it out of the text visible to the user. In other words, changing `foo@example.com` to `<mailto:foo@example.com>` is no good because it changes the text as seen by the user. However, `[foo@example.com](mailto:foo@example.com)` is fine.

## What I could do to make this happen

These seem like things that will work, but are (to me, at least) not as ideal as what is suggested in the previous section.

* Write my own plugin to change `foo@example.com` to `[foo@example.com](mailto:foo@example.com)` and have it do its transformations before `remark-gfm`.
* Disable `remark-lint-no-auto-link-without-protocol`.

## One thing that's puzzling me

`remark-lint-no-auto-link-without-protocol` docs state that [it can be used to fix lint violations](https://github.com/remarkjs/remark-lint/tree/main/packages/remark-lint-no-auto-link-without-protocol#fix), but I have not been unable to make that work. If it does work (and I'm just using it wrong or something), then that is likely an ideal solution. But I'm not seeing how to make it work, at least not without writing my own code.

`input.md`:

```markdown
This needs a mailto protocol: <tituswormer@example.com>
```

```console
$ remark -u lint-no-auto-link-without-protocol input.md -o output.md
input.md > output.md
  1:31-1:56  warning  All automatic links must start with a protocol  no-auto-link-without-protocol  remark-lint

⚠ 1 warning
$ cat output.md
This needs a mailto protocol: <tituswormer@example.com>
$ 
```

Maybe I need to pass a setting with `-s`?

**Comment by wooorm** [comment, marked as answer]
I believe historically there was a lot of difference between how “pedantic” markdown (Gruber-style), old GFM (from before CM), and CM, parsed these.
I think the lint rule being weird is a remnant of that, which can be revisited:

```markdown
a@b.c

<a@b.c>

<mailto:a@b.c>

[a@b.c](mailto:a@b.c)
```

a@b.c

<a@b.c>

<mailto:a@b.c>

[a@b.c](mailto:a@b.c)

**Reply by wooorm** [reply]
I guess the whole rule is useless now. It’s no longer possible to start an autolink (other than an email one), without a protocol. And it’s wrong about email links. 🤔 

I think it should be deprecated.

**Reply by Trott** [reply]
If that's the case, then I guess disabling the rule in my configuration is both low effort and an excellent option for me.

**Reply by wooorm** [reply]
I’ve removed the package in question from the presets!
